language/bytecode_verifier/src/type_memory_safety.rs

```diff
@@ -221,63 +221,105 @@ impl<'a> TypeAndMemorySafetyAnalysis<'a> {
-                    if self
-                        .module
-                        .is_field_in_struct(*field_definition_index, struct_handle_index)
-                    {
-                        let field_signature = self
-                            .module
-                            .get_field_signature(*field_definition_index)
-                            .0
-                            .clone();
-                        let operand_nonce = operand.value.extract_nonce().unwrap().clone();
-                        let nonce = self.get_nonce(&mut state);
-                        if operand.signature.is_mutable_reference() {
-                            let borrowed_nonces = state.borrowed_nonces_for_field(
-                                *field_definition_index,
-                                operand_nonce.clone(),
-                            );
-                            if Self::write_borrow_ok(borrowed_nonces) {
-                                self.stack.push(StackAbstractValue {
-                                    signature: SignatureToken::MutableReference(Box::new(
-                                        field_signature.clone(),
-                                    )),
-                                    value: AbstractValue::Reference(nonce.clone()),
-                                });
-                                state.borrow_field_from_nonce(
-                                    *field_definition_index,
-                                    operand_nonce.clone(),
-                                    nonce,
-                                );
-                                state.destroy_nonce(operand_nonce);
-                            } else {
-                                return Err(
-                                    VMStaticViolation::BorrowFieldExistsMutableBorrowError(offset),
-                                );
-                            }
-                        } else {
-                            self.stack.push(StackAbstractValue {
-                                signature: SignatureToken::Reference(Box::new(field_signature)),
-                                value: AbstractValue::Reference(nonce.clone()),
-                            });
-                            state.borrow_field_from_nonce(
-                                *field_definition_index,
-                                operand_nonce.clone(),
-                                nonce,
-                            );
-                            state.destroy_nonce(operand_nonce);
+                    // TODO Separate Mut and Imm BorrowField Errors
+                    return Err(VMStaticViolation::BorrowFieldBadFieldError(offset));
+                }
+
+                let field_signature = self
+                    .module
+                    .get_field_signature(*field_definition_index)
+                    .0
+                    .clone();
+                let operand_nonce = operand.value.extract_nonce().unwrap().clone();
```

Comment:
`clone` needed?

Reply:
I think all of these clones are due to references being used

Reply:
Ah, nevermind. the operand one is necessary, but not the signature one